AnnService/src/Helper/AsyncFileReader.cpp

```diff
@@ -82,39 +82,48 @@ namespace SPTAG {
             }
         }
 #else
-        void BatchReadFileAsync(std::vector<std::shared_ptr<Helper::DiskPriorityIO>>& handlers, AsyncReadRequest* readRequests, int num)
+        using BatchOp = bool (DiskPriorityIO::*)(AsyncReadRequest*, uint32_t);
+        void CallOnAppropriateBatch(std::vector<std::shared_ptr<Helper::DiskPriorityIO>>& handlers, AsyncReadRequest* readRequests, int num, BatchOp f)
         {
-            
             if (handlers.size() == 1) {
-                handlers[0]->BatchReadFile(readRequests, num);
+                (handlers[0].get()->*f)(readRequests, num);
             }
-            else {            
+            else {
                 int currFileId = 0, currReqStart = 0;
                 for (int i = 0; i < num; i++) {
                     AsyncReadRequest* readRequest = &(readRequests[i]);
                     if (readRequest->m_readSize == 0) continue;
 
                     int fileid = (readRequest->m_status >> 16);
                     if (fileid != currFileId) {
-                        handlers[currFileId]->BatchReadFile(readRequests + currReqStart, i - currReqStart);
+                        (handlers[currFileId].get()->*f)(readRequests + currReqStart, i - currReqStart);
                         currFileId = fileid;
                         currReqStart = i;
                     }
                 }
                 if (currReqStart < num) {
-                    handlers[currFileId]->BatchReadFile(readRequests + currReqStart, num - currReqStart);
+                    (handlers[currFileId].get()->*f)(readRequests + currReqStart, num - currReqStart);
                 }
             }
+        }
+
+        void BatchReadFileAsync(std::vector<std::shared_ptr<Helper::DiskPriorityIO>>& handlers, AsyncReadRequest* readRequests, int num)
+        {
+            
+            CallOnAppropriateBatch(handlers, readRequests, num, &DiskPriorityIO::BatchReadFile);
 
             for (int i = 0; i < num; i++) {
                 AsyncReadRequest* readRequest = &(readRequests[i]);
                 if (readRequest->m_readSize == 0) continue;
-                readRequest->m_readSize = 0;
-                if (readRequest->m_success)
+                
+                if (readRequest->m_success && readRequest->m_callback)
```

Comment:
If we use the readSize to judge whether the request is available then we do not need to test the m_callback? Or we change the logic to use m_callback to test the availability?

Reply:
With the cleanup, there needs to be an extra state where the read has been process, but the resource has not been cleaned up.

In this implementation, we use the m_callback to check if it has been processed, and then the readSize to judge if the resource has been cleaned up

Reply:
Then we need to do the line 112 after the cleanup to make sure that the readSize is cleaned after the request is executed since when a posting contains 0 items, it will be skipped in the ExtraSearcher without setting the readSize to 0.

Reply:
I move the skip before the m_extraSearcher->SearchIndex call - it handles this case, and also means we will not sacrifice a read if one of the postings contains 0 items but we have more than `searchInternalResultNum` candidates from the head